atomicfu-transformer/src/main/kotlin/kotlinx/atomicfu/transformer/AtomicFUTransformer.kt

```diff
@@ -785,33 +775,71 @@ class AtomicFUTransformer(
 
         private fun fixupLoadedAtomicVar(f: FieldInfo, ld: FieldInsnNode): AbstractInsnNode? {
             val j = FlowAnalyzer(ld.next).execute()
-            when (j) {
+            return fixupOperationOnAtomicVar(j, f, ld, null)
+        }
+
+        private fun fixupLoadedArrayElement(f: FieldInfo, ld: FieldInsnNode, getter: MethodInsnNode): AbstractInsnNode? {
+            // contains array field load (in vh case: + swap and pure type array load) and array element index
+            val arrayElementInfo = mutableListOf<AbstractInsnNode>()
+            if (vh) {
+                arrayElementInfo.add(ld.previous.previous) // getstatic VarHandle field
+                arrayElementInfo.add(ld.previous) // swap
+            }
+            var i: AbstractInsnNode = ld
+            while (i != getter) {
+                arrayElementInfo.add(i)
```

Comment:
This is very brittle. It will not work with complex expression in array access, like `arr[<complex-expression>].update { ... }`. What's worse, with code code `arr[funCall()].update { ... }`, the code for `funCall()` will get duplicated and code semantics will change.

Please, add more diagnostics here. Similarly to data-flow analyzer, please ensure that this piece of code that is being stored in `arrayElementInfo` does not contain any complex data-flow. Moreover, it should not contain any invocations. Only simple loads and simple arithmetics can be allowed here. If anything else is encountered, then please complain with an appropriate message.